File: 0200_number_of_islands/number_of_islands.py
from typing import List


def num_islands(grid: List[List[str]]) -> int:
    """
    Time: O(m * n)
    Space: O(m * n)
    """
    if not len(grid):
        return 0
    islands = 0
    for i in range(len(grid)):
        for j in range(len(grid[0])):
            if grid[i][j] == "1":
                islands += dfs(grid, i, j)
    return islands


def dfs(grid: List[List[str]], row: int, col: int) -> int:
    if (row < 0 or col < 0 or row >= len(grid) or col >= len(grid[0])
            or grid[row][col] == "0"):
        return
    grid[row][col] = "0"
    dfs(grid, row + 1, col)
    dfs(grid, row, col + 1)
    dfs(grid, row - 1, col)
    dfs(grid, row, col - 1)
    return 1

File: 0200_number_of_islands/number_of_islands.cpp
#include <iostream>
#include <vector>

using namespace std;

int dfs(vector<vector<char>> &grid, int row, int col)
{
    if (row < 0 || col < 0 || row >= grid.size() ||
        col >= grid[0].size() || grid[row][col] == '0')
        return 0;
    grid[row][col] = '0';
    dfs(grid, row - 1, col);
    dfs(grid, row, col - 1);
    dfs(grid, row + 1, col);
    dfs(grid, row, col + 1);
    return 1;
}

/**
 * Time: O(m * n)
 * Space: O(m * n)
 */
int num_islands(vector<vector<char>> &grid)
{
    int islands = 0;
    for (int i = 0; i < grid.size(); i++)
        for (int j = 0; j < grid[0].size(); j++)
            if (grid[i][j] == '1')
                islands += dfs(grid, i, j);
    return islands;
}
